How many late days are you using for this assignment? 1

In [173]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [174]:
# Load the given datasets
numerical = pd.read_csv("/Users/<redacted>/Downloads/chronic_kidney_disease_numerical.csv")
categorical = pd.read_csv("/Users/<redacted>/Downloads/chronic_kidney_disease_categorical.csv")

# Print the data
print(numerical)
print(categorical)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The Chronic Kidney Disease dataset can be used to predict the chronic kidney disease and it can be collected from the hospital nearly 2 months of period. The features are: age, bp, sg, al, su, rbc, pc, pcc, ba, bgr, bu, sc, sod, pot, hemo, pcv, wc, rc, htn, dm, cad, appet, pe, and ane. The labels are ckd and notckd.

The features in the given datasets are both categorical and numerical. 
Examples for categorical: al, su, rbc.
Examples for numerical: age, bp, bgr

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [175]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print(numerical.duplicated().sum())

# Drop these duplicate rows
drop_dup_numerical = numerical.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(categorical.duplicated().sum())
print(categorical[categorical.duplicated()])
drop_dup_categorical = categorical.drop_duplicates()


30
30
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
231     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
401     628271  2.0  1.0  abnormal    normal  notpresent  notpresent  yes   
402     158929  1.0  0.0       NaN       NaN  notpresent  notpresent   no   
403     240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
404     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
405     337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
406     118191  0.0  0.0    normal    normal  notpresent  notpresent   no   
407     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
408     302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
409     938027  0.0  0.0    normal    normal  notpresent  notpresent   no   
410     572140  0.0  0.0    normal    normal  notpresent  notpresent   no   
411     283427  0.0  0.0    normal    normal  notpresent  notpresent  

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
No, not all unique ids are present in both datasets. The size of the merged dataet is 400x25, which is clearly less than what we had previously. This is because there are unique IDs in the numerical dataset that are not present in the categorical dataset. Additionally, the union join is similar to a union of the datasets and takes the rows of both dataset irrespective of whether they are common in both.

In [176]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_dataset = pd.merge(drop_dup_numerical, drop_dup_categorical, how = "outer" , on="unique_id")

#Print the combined dataset
print(merged_dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [177]:
# Calculate the percentage of rows that contain atleast one missing value
rows_na = merged_dataset.isnull().any(axis=1).sum()
total_rows = merged_dataset.shape[0]
percentage_na = (rows_na/total_rows)*100

# Print %
print(percentage_na)

# Drop these rows from the dataset
merged_dataset_cleaned = merged_dataset.dropna()

# Print the Dataset
print(merged_dataset_cleaned)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc        

## Part 3.4: Sort the dataset according to the Labels

In [178]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_dataset = merged_dataset_cleaned.sort_values(by = 'Target').reset_index(drop = True)

# Print the dataset
print(sorted_dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [179]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_dataset = pd.get_dummies(sorted_dataset, columns = categorical_columns, drop_first = True, dtype = int)

# Print the dataset
print(encoded_dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 
When a binary variable has only two categories, representing it with a single column (e.g., encoding rbc as 0 for abnormal and 1 for normal) is similar to One-Hot Encoding. This is because all necessary information can be effectively conveyed in just one column.
However, if the categorical variable has more than two categories, using a single column with integer labels (like 0, 1, 2, 3, 4) may create an unintended ordinal relationship. For instance, if we have a "type of condiment" variable with five different categories, this approach could imply a ranking that does not actually exist among these categories.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [180]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers(df, numerical_columns):
    for column in numerical_columns:
        mean = df[column].mean()
        std_dev = df[column].std()
        lower_limit = mean - 3 * std_dev
        upper_limit = mean + 3 * std_dev
        df = df[(df[column] >= lower_limit) & (df[column] <= upper_limit)]
    return df

cleaned_dataset = remove_outliers(encoded_dataset, numerical_columns)

# Print the dataset
print(cleaned_dataset)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [181]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = cleaned_dataset.select_dtypes(include = [int, float]).columns
scaler = MinMaxScaler()
cleaned_dataset[numerical_columns] = scaler.fit_transform(cleaned_dataset[numerical_columns])

# Print the dataset
print(cleaned_dataset)

     unique_id       age    bp       bgr        bu        sc       sod  \
2     0.264544  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5     0.325059  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7     0.371049  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8     0.411888  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9     0.412782  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
148   0.425405  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
149   0.106438  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
150   0.420566  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
151   0.135054  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
152   1.000000  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
2    0.793103  0.000000  0.032258

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:
Yes, the column in this dataset which is not appropriate for modeling and predictions is unique_id. This column is not necessary because it does not have a direct impact on a person's health and is unnecessary to the dataset.

In [182]:
#Remove that column
final = cleaned_dataset.drop(columns = ['unique_id'])
# Print the dataset
print(final)

          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
2    0.000000  0.032258  0.395161  ...       

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [183]:
# Export the dataframe to a new csv file
final.to_csv('final.csv', index = True)
# Export the dataframe to a new json file
final.to_json('final.json', orient = 'records', lines = True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell below

Paste here: 

|   age   |   bp  |   bgr   |   bu   |   sc   |   sod   |   pot   |   hemo   |   pcv   |   wbcc   |   ...   |   pc_normal  |   pcc_present  |   ba_present |   htn_yes |   dm_yes |   cad_yes |   appet_poor |   pe_yes |   ane_yes |   Target_notckd |
|---------|-------|---------|--------|--------|---------|---------|----------|---------|----------|---------|--------------|----------------|--------------|-----------|----------|-----------|---------------|----------|-----------|------------------|
| 0.743243| 0.50  | 0.442060| 0.901961| 0.432099| 0.500000| 0.793103| 0.000000 | 0.032258| 0.395161 |   ...   | 0.0          | 0.0            | 0.0          | 1.0       | 1.0      | 1.0       | 1.0          | 1.0      | 1.0       | 0.0              |
| 0.770270| 1.00  | 0.901288| 0.163399| 0.345679| 0.766667| 0.206897| 0.524752 | 0.548387| 0.443548 |   ...   | 1.0          | 0.0            | 1.0          | 1.0       | 0.0      | 1.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.905405| 0.50  | 0.785408| 0.862745| 0.518519| 0.600000| 1.000000| 0.277228 | 0.322581| 0.233871 |   ...   | 0.0          | 0.0            | 0.0          | 1.0       | 1.0      | 1.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.202703| 0.75  | 0.158798| 0.196078| 0.160494| 0.166667| 0.206897| 0.059406 | 0.000000| 0.653226 |   ...   | 0.0          | 1.0            | 1.0          | 0.0       | 0.0      | 0.0       | 0.0          | 0.0      | 1.0       | 0.0              |
| 0.783784| 1.00  | 0.399142| 0.287582| 0.839506| 0.666667| 0.586207| 0.019802 | 0.096774| 0.258065 |   ...   | 0.0          | 0.0            | 1.0          | 1.0       | 1.0      | 0.0       | 1.0          | 0.0      | 0.0       | 0.0              |
| 0.729730| 0.00  | 0.935622| 0.169935| 0.160494| 0.333333| 0.034483| 0.019802 | 0.064516| 0.879032 |   ...   | 0.0          | 1.0            | 0.0          | 1.0       | 0.0      | 1.0       | 0.0          | 0.0      | 1.0       | 0.0              |
| 0.675676| 0.75  | 0.253219| 0.633987| 0.777778| 0.366667| 0.655172| 0.138614 | 0.193548| 0.169355 |   ...   | 0.0          | 0.0            | 0.0          | 1.0       | 0.0      | 0.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.851351| 0.25  | 0.618026| 0.562092| 0.728395| 0.000000| 0.344828| 0.168317 | 0.161290| 0.580645 |   ...   | 0.0          | 1.0            | 1.0          | 1.0       | 1.0      | 1.0       | 0.0          | 1.0      | 1.0       | 0.0              |
| 0.540541| 0.00  | 0.399142| 0.535948| 0.358025| 0.700000| 0.379310| 0.207921 | 0.161290| 0.830645 |   ...   | 1.0          | 0.0            | 0.0          | 1.0       | 1.0      | 0.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.756757| 0.25  | 0.223176| 0.209150| 0.160494| 0.533333| 0.620690| 0.485149 | 0.516129| 0.290323 |   ...   | 0.0          | 0.0            | 0.0          | 1.0       | 1.0      | 0.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.662162| 0.50  | 0.618026| 0.411765| 0.432099| 0.566667| 0.689655| 0.316832 | 0.354839| 0.250000 |   ...   | 0.0          | 1.0            | 1.0          | 1.0       | 1.0      | 0.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.783784| 0.00  | 0.725322| 0.313725| 0.481481| 0.566667| 0.862069| 0.178218 | 0.193548| 0.258065 |   ...   | 0.0          | 0.0            | 1.0          | 1.0       | 1.0      | 0.0       | 1.0          | 1.0      | 0.0       | 0.0              |
| 0.878378| 0.00  | 0.206009| 0.751634| 0.604938| 0.533333| 0.689655| 0.366337 | 0.387097| 0.879032 |   ...   | 1.0          | 0.0            | 0.0          | 1.0       | 1.0      | 0.0       | 1.0          | 1.0      | 0.0       | 0.0              |
| 0.878378| 0.25  | 0.639485| 0.470588| 0.395062| 0.433333| 0.517241| 0.267327 | 0.322581| 0.104839 |   ...   | 0.0          | 1.0            | 1.0          | 1.0       | 1.0      | 0.0       | 0.0          | 0.0      | 0.0       | 0.0              |
| 0.716216| 0.50  | 1.000000| 0.163399| 0.111111| 0.066667| 0.206897| 0.267327 | 0.387097| 0.532258 |   ...   | 1.0          | 0.0            | 0.0          | 1.0       | 0.0      | 0.0       | 0.0          | 0.0      | 0.0       | 0.0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

To mitigate hallucinations during LLM data conversions, use clear specifications and structured prompts. Break the process into smaller steps and validate outputs against known data sources. Involve human oversight for critical tasks and iterate on prompts based on observed errors. Limit ambiguity in instructions and test the model with edge cases. Maintain logs to analyze errors and refine approaches. For domain-specific tasks, consider fine-tuning the model on relevant datasets. These strategies enhance accuracy and reliability in data conversions.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [184]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'unique_id': 'Unique ID',
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_normal': 'Red_Blood_Cells_normal',
    'pc_normal': 'Pus_Cell_normal',
    'pcc_present': 'Pus_Cell_Clumps_present',
    'ba_present': 'Bacteria_present',
    'htn_yes': 'Hypertension_yes',
    'dm_yes': 'Diabetes_Mellitus_yes',
    'cad_yes': 'Coronary_Artery_Disease_yes',
    'appet_poor': 'Appetite_poor',
    'pe_yes': 'Pedal_Edema_yes',
    'ane_yes': 'Anemia_yes',
    'Target_notckd': 'Target_notckd'
}

final.rename(columns = rename_mapping, inplace = True)

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [185]:
# Converted SQL to Pandas code
result = final.groupby('Target_notckd')['Target_notckd'].count().reset_index(name='count')
result.rename(columns = {'Target_notckd': 'Target'}, inplace = True)
print(result)

   Target  count
0     0.0     22
1     1.0    111


The above code groups the dataframe by the values in the column Target_notckd. After grouping, the number of occurrences of each unique value in the column is counted and printed in the column renamed as Target. The target 0 values represent not_ckd and 1 values represent ckd.